python/cudf/cudf/core/join/join.py

```diff
@@ -255,20 +255,9 @@ def perform_merge(self) -> cudf.DataFrame:
         for left_key, right_key in zip(self._left_keys, self._right_keys):
             lcol = left_key.get(self.lhs)
             rcol = right_key.get(self.rhs)
-            lcol_casted, rcol_casted = _match_join_keys(lcol, rcol, self.how)
+            lcol_casted, rcol_casted = match_join_types(lcol, rcol)
             left_join_cols.append(lcol_casted)
             right_join_cols.append(rcol_casted)
-
-            # Categorical dtypes must be cast back from the underlying codes
-            # type that was returned by _match_join_keys.
-            if (
-                self.how == "inner"
-                and isinstance(lcol.dtype, cudf.CategoricalDtype)
-                and isinstance(rcol.dtype, cudf.CategoricalDtype)
-            ):
```

Comment:
Pandas never did this, AFAICT. For an inner merge if the _left_ column was categorical, it would decat the left column and convert the right column to object (if not categorical), and vice versa. The resulting key column in the output would be from the left column. So if you were merging on (cat, int) you would get an `int` key column, and on (int, cat) you get an `object` key column:

```
n [2]: right = pd.DataFrame({"key": pd.Series([1, 2], dtype="category"), "value": [2, 1]})

In [3]: left = pd.DataFrame({"key": [1, 3], "value": [1, 2]})

In [4]: left.merge(right, on="key", how="inner").dtypes
Out[4]: 
key        object
value_x     int64
value_y     int64
dtype: object

In [5]: left.merge(right, on="key", how="inner")
Out[5]: 
  key  value_x  value_y
0   1        1        2

In [6]: right.merge(left, on="key", how="inner").dtypes
Out[6]: 
key        int64
value_x    int64
value_y    int64
dtype: object

In [7]: left.merge(right, on="key", how="inner")
Out[7]: 
  key  value_x  value_y
0   1        1        2

# If both are cat, then we decat and merge.
In [8]: left = pd.DataFrame({"key": pd.Series([1, 3], dtype="category"), "value": [1, 2]})

In [9]: left.merge(right, on="key", how="inner")
Out[9]: 
   key  value_x  value_y
0    1        1        2

In [10]: left.merge(right, on="key", how="inner").dtypes
Out[10]: 
key        int64
value_x    int64
value_y    int64
dtype: object

```